### PR TITLE
🌱 Feature: Add scorecard attestation policy module

### PR DESCRIPTION
**Pull request by raghavkaul**
#### What kind of change does this PR introduce?

Add submodule for attestation policy in scorecard, as per #2136.

#### Preliminary features added

* Policy parsing
* Policy evaluation for Binary-Artifacts, Vulnerabilities, Pinned-Dependencies and Code-Review checks

- [x] Tests for the changes have been added (for bug fixes/features)

#### Special notes for your reviewer
This is the first changeset in a series that will attempt to implement #2136. Future changesets will add a user-facing option to use the attestation module, but currently, this is code is not reachable.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: checks/evaluation/pinned_dependencies.go [hunk 1]
```diff
@@ -26,19 +26,19 @@ import (
...
 )
 
 // Structure to host information about pinned github
 // or third party dependencies.
-type worklowPinningResult struct {
-	thirdParties pinnedResult
-	gitHubOwned  pinnedResult
+type WorkflowPinningResult struct {
+	ThirdParties PinnedResult
+	GitHubOwned  PinnedResult
```
Comment (azeemshaikh38) [inline checks/evaluation/pinned_dependencies.go:41]: These need not be exposed. `evaluation` package is not something the attestor should depend on. It should work on the raw data directly - https://github.com/ossf/scorecard/blob/main/checker/raw_result.go#L26.

File: policy/attestation_policy.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	// of vulnerabilities, as discovered from the OSV service
+	EnsureNoVulnerabilities bool `yaml:"ensureNoVulnerabilities"`
+
+	// EnsurePinnedDependencies : set to true to require that this project pin dependencies
+	// by hash/commit SHA
+	EnsurePinnedDependencies bool `yaml:"ensurePinnedDependencies"`
+
+	// EnsureCodeReviewed : set to true to require that the most recent commits in
+	// this project have gone through a code review process
+	EnsureCodeReviewed bool `yaml:"ensureCodeReviewed"`
```
Comment (azeemshaikh38) [inline policy/attestation_policy.go:50]: Shouldn't these be more granular? For example, how can a user ignore pinned dependencies for a specific file or for a specific library? For Code-Review, lets also expose options to set `minApprovals` and `requiredApprovers`?

**Review by azeemshaikh38 — COMMENTED** [review]
Let's add the new code to a new directory (either `attestor` or `signer`) in the root of the repository. This should be a separate go module and not a package.

**Review by laurentsimon — APPROVED** [review]
LGTM. Do you want @jeffmendoza to take a look?

**Review by azeemshaikh38 — APPROVED** [review]
LGTM with comments

**Comment by jeffmendoza** [comment]
> LGTM. Do you want @jeffmendoza to take a look?

LGTM. Looks to me in the direction of what Allstar can utilize 👍

**Comment by azeemshaikh38** [comment]
@raghavkaul thanks merged. Just fyi that if you have addressed the PR comments, feel free to mark them as resolved

**Comment by spencerschrock** [comment]
I'm getting e2e failures after this got merged:

```
==================
WARNING: DATA RACE
Read at 0x00c000382528 by goroutine 16:
  github.com/ossf/scorecard/v4/utests.(*TestDetailLogger).Info()
      .../go/pkg/mod/github.com/ossf/scorecard/v4@v4.6.0/utests/utlib.go:49 +0xc4
  github.com/ossf/scorecard-attestor.CheckPreventBinaryArtifacts()
      .../go/src/github.com/spencerschrock/scorecard.git/attestor/attestation_policy.go:78 +0x41a
  github.com/ossf/scorecard-attestor.TestCheckPreventBinaryArtifacts.func1()
      .../go/src/github.com/spencerschrock/scorecard.git/attestor/attestation_policy_test.go:110 +0xcd
  testing.tRunner()
      .../src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      .../src/testing/testing.go:1493 +0x47

Previous write at 0x00c000382528 by goroutine 17:
  github.com/ossf/scorecard/v4/utests.(*TestDetailLogger).Info()
      .../go/pkg/mod/github.com/ossf/scorecard/v4@v4.6.0/utests/utlib.go:49 +0x197
  github.com/ossf/scorecard-attestor.CheckPreventBinaryArtifacts()
      .../go/src/github.com/spencerschrock/scorecard.git/attestor/attestation_policy.go:78 +0x41a
  github.com/ossf/scorecard-attestor.TestCheckPreventBinaryArtifacts.func1()
      .../go/src/github.com/spencerschrock/scorecard.git/attestor/attestation_policy_test.go:110 +0xcd
  testing.tRunner()
      .../src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      .../src/testing/testing.go:1493 +0x47

Goroutine 16 (running) created at:
  testing.(*T).Run()
      .../src/testing/testing.go:1493 +0x75d
  github.com/ossf/scorecard-attestor.TestCheckPreventBinaryArtifacts()
      .../go/src/github.com/spencerschrock/scorecard.git/attestor/attestation_policy_test.go:108 +0xd24
=== CONT  TestCheckPreventBinaryArtifacts/test_with_multiple_ignored_artifacts
  testing.tRunner()
      .../src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      .../src/testing/testing.go:1493 +0x47

Goroutine 17 (finished) created at:
  testing.(*T).Run()
      .../src/testing/testing.go:1493 +0x75d
  github.com/ossf/scorecard-attestor.TestCheckPreventBinaryArtifacts()
      .../go/src/github.com/spencerschrock/scorecard.git/attestor/attestation_policy_test.go:108 +0xd24
  testing.tRunner()
      .../src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      .../src/testing/testing.go:1493 +0x47
==================
=== CONT  TestCheckPreventBinaryArtifacts/test_with_some_artifacts
    testing.go:1319: race detected during execution of test
--- FAIL: TestAttestationPolicyRead (0.00s)
    --- PASS: TestAttestationPolicyRead/default_attestation_policy_with_everything_on (0.00s)
    --- PASS: TestAttestationPolicyRead/policy_with_allowlist_of_binary_artifacts#01 (0.00s)
    --- PASS: TestAttestationPolicyRead/policy_with_allowlist_of_binary_artifacts (0.00s)
    --- PASS: TestAttestationPolicyRead/invalid_attestation_policy (0.00s)
--- FAIL: TestCheckPreventBinaryArtifacts (0.00s)
    --- PASS: TestCheckPreventBinaryArtifacts/test_with_no_artifacts (0.00s)
    --- PASS: TestCheckPreventBinaryArtifacts/test_with_multiple_artifacts (0.00s)
    --- PASS: TestCheckPreventBinaryArtifacts/test_with_glob_ignored (0.00s)
    --- PASS: TestCheckPreventBinaryArtifacts/test_with_multiple_ignored_artifacts (0.00s)
    --- FAIL: TestCheckPreventBinaryArtifacts/test_with_some_artifacts (0.01s)
FAIL
coverage: 77.4% of statements
ginkgo run failed
  could not finalize profiles:
  Could not process Coverprofile e2e-coverage.out: exit status 1
  ```

**Pull request merged**